In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_excel('Test.xlsx')
data.columns = data.columns.str.strip()

In [3]:
pd.set_option('display.max_columns', None)

In [4]:
data.describe()

Date  Opening Weights  Open Quantity  \
count                           7721      7721.000000    7721.000000   
mean   2022-07-24 21:10:39.243621376         0.012899    3068.103354   
min              2022-01-03 00:00:00        -0.857720 -575000.000000   
25%              2022-05-02 00:00:00        -0.019858    -571.000000   
50%              2022-08-09 00:00:00         0.019740     184.000000   
75%              2022-10-21 00:00:00         0.043347    1690.000000   
max              2022-12-29 00:00:00         1.021280  690000.000000   
std                              NaN         0.088113   38129.677302   

       Close Quantity  Closing Weights        Price  Exchange Rate  \
count     7721.000000      7721.000000  7721.000000    7721.000000   
mean      3076.421837         0.013114   139.493400       0.912491   
min    -575000.000000        -0.840360     0.348000       0.007202   
25%       -571.000000        -0.019996    14.530000       1.000000   
50%        185.000000         0.019647    42.430000       1.000000   
75%       1718.000000         0.043359   120.525002       1.000000   
max     690000.000000         1.050469  4798.250000       1.145050   
std      38137.989546         0.088608   338.412167       0.282734   

       Value in USD  Price Yesterday  Stock Movement  \
count  7.721000e+03      7721.000000     7721.000000   
mean   1.649024e+04       139.658694       -0.001052   
min   -1.058781e+06         0.348000       -0.429210   
25%   -2.533216e+04        14.530000       -0.017148   
50%    2.366320e+04        42.540001        0.000000   
75%    5.309280e+04       120.959999        0.014862   
max    1.356712e+06      4920.750000        0.340782   
std    1.163459e+05       338.914054        0.038409   

       Performance Contribution   Traded Today  Trade Price  Trade Day Move  \
count               7721.000000    1644.000000  1644.000000     1644.000000   
mean                   0.000039      39.067518   149.721297        0.001144   
min                   -0.020470 -600000.000000     0.382376       -0.273464   
25%                   -0.000420    -861.250000    22.281586       -0.010627   
50%                    0.000000       2.000000    54.609850       -0.000085   
75%                    0.000430     865.250000   145.330167        0.010542   
max                    0.036137  600000.000000  4848.548979        0.364672   
std                    0.001676   23812.621085   334.435418        0.034355   

       Trade Weight  Return Adjustments  Total Return  Calculated NAV  \
count   1644.000000         1644.000000   7721.000000    7.721000e+03   
mean      -0.000125            0.000094      0.000059    1.253228e+06   
min       -1.016014           -0.023953     -0.023953    8.802463e+05   
25%       -0.029442           -0.000250     -0.000494    1.167917e+06   
50%        0.000230            0.000013      0.000000    1.297075e+06   
75%        0.029398            0.000332      0.000500    1.363642e+06   
max        0.973446            0.035505      0.036137    1.540782e+06   
std        0.095824            0.002119      0.001890    1.536410e+05   

       NAV Yesterday  Close Weight ABS    Dollar PnL     SharesOut  \
count   7.721000e+03       7721.000000   7721.000000  7.721000e+03   
mean    1.254114e+06          0.046311     62.515332  4.898625e+06   
min     8.753504e+05          0.000000 -35086.342986  1.437230e+00   
25%     1.164112e+06          0.019771   -613.079041  1.258000e+02   
50%     1.308857e+06          0.031259      0.000000  4.428000e+02   
75%     1.364549e+06          0.050182    615.866562  3.584462e+03   
max     1.490786e+06          1.050469  45916.036818  3.904580e+08   
std     1.564948e+05          0.076671   2345.653395  3.084622e+07   

         Market_Cap  
count  7.721000e+03  
mean   2.683242e+08  
min    2.208729e+02  
25%    3.495043e+03  
50%    1.394949e+04  
75%    8.155872e+04  
max    1.463069e+10  
std    1.683322e+09

In [5]:
data.head()  

Date P_Ticker        Short_Name  Opening Weights  Open Quantity  \
0 2022-01-03     ADBE         ADOBE INC         0.023996             39   
1 2022-01-03      ADE  Bitcoin Group SE         0.019372            387   
2 2022-01-03      ADI    ANALOG DEVICES         0.053208            279   
3 2022-01-03     AMAT  APPLIED MATERIAL         0.023723            139   
4 2022-01-03      AMC  AMC ENTERTAINMEN        -0.006610           -224   

   Close Quantity  Closing Weights Currency       Price  Exchange Rate  \
0              39         0.023680      USD  564.250000         1.0000   
1             387         0.019406   EURUSD   41.250000         1.1297   
2             279         0.053174      USD  177.115005         1.0000   
3             139         0.023920      USD  159.925003         1.0000   
4            -224        -0.006394      USD   26.525000         1.0000   

   Value in USD  Price Yesterday  Stock Movement  Performance Contribution  \
0  22005.750000       567.239990       -0.005271                 -0.000126   
1  18034.247504        40.849998        0.009792                  0.000190   
2  49415.086533       175.820007        0.007365                  0.000392   
3  22229.575424       157.345001        0.016397                  0.000389   
4  -5941.599915        27.205000       -0.024995                  0.000165   

   Traded Today  Trade Price  Trade Day Move  Trade Weight  \
0           NaN          NaN             NaN           NaN   
1           NaN          NaN             NaN           NaN   
2           NaN          NaN             NaN           NaN   
3           NaN          NaN             NaN           NaN   
4           NaN          NaN             NaN           NaN   

   Return Adjustments  Total Return  Calculated NAV  NAV Yesterday Country  \
0                 NaN     -0.000126   929315.622804    921929.6875      US   
1                 NaN      0.000190   929315.622804    921929.6875      DE   
2                 NaN      0.000392   929315.622804    921929.6875      US   
3                 NaN      0.000389   929315.622804    921929.6875      US   
4                 NaN      0.000165   929315.622804    921929.6875      US   

                   Sector                        Industry  \
0  Information Technology                        Software   
1              Financials                 Capital Markets   
2  Information Technology  Semiconductors & Semiconductor   
3  Information Technology  Semiconductors & Semiconductor   
4  Communication Services                   Entertainment   

                   Sub_Ind  IS_Currency  Close Weight ABS  IS_Future  \
0     Application Software        False          0.023680      False   
1                   Crypto        False          0.019406      False   
2           Semiconductors        False          0.053174      False   
3  Semiconductor Equipment        False          0.023920      False   
4   Movies & Entertainment        False          0.006394      False   

   Short_Pos  Dollar PnL  SharesOut     Market_Cap Cap Class  
0      False -116.609619  448.00000  252784.000000      Mega  
1      False  174.878219    5.00000     233.000614     Micro  
2      False  361.304489  495.90815   87832.774711     Large  
3      False  358.620255  830.00000  132737.752533     Large  
4       True  152.320068  295.58775    7840.464956      SMID

### Check Exchange Rate Ranges

This cell defines a function `check_exchange_rate_ranges` to validate whether the exchange rates for different currencies fall within predefined reasonable ranges. It flags rows where the exchange rate is outside the expected range and returns a DataFrame containing the flagged rows. 

#### Formula:
$$
\text{Exchange Rate Error} = 
\begin{cases} 
\text{True}, & \text{if } \text{Exchange Rate} \notin [\text{Min Range}, \text{Max Range}] \\
\text{False}, & \text{otherwise}
\end{cases}
$$


In [6]:
def check_exchange_rate_ranges(data):
    currency_ranges = {
        'USD': (1, 1),
        'EURUSD': (0.95, 1.2),
        'GBPUSD': (1.20, 1.45),
        'HKDUSD': (0.127, 0.129),
        'CADUSD': (0.69, 0.83),
        'SEKUSD': (0.087, 0.125),
        'CHFUSD': (0.91, 1.18),
        'JPYUSD': (0.0063, 0.0111),
        'PLNUSD': (0.22, 0.31)
    }
    
    data['Exchange Rate Error'] = data.apply(
        lambda row: not (currency_ranges[row['Currency']][0] <= row['Exchange Rate'] <= currency_ranges[row['Currency']][1])
        if row['Currency'] in currency_ranges else False,
        axis=1
    )
    exchange_rate_errors = data[data['Exchange Rate Error']]
    
    return exchange_rate_errors[['Date', 'P_Ticker', 'Currency', 'Exchange Rate', 'Exchange Rate Error']]

# Example usage
exchange_rate_errors = check_exchange_rate_ranges(data)
exchange_rate_errors

Date P_Ticker Currency  Exchange Rate  Exchange Rate Error
6    2022-01-03     ANTO   GBPUSD       0.013480                 True
27   2022-01-04     ANTO   GBPUSD       0.013527                 True
46   2022-01-05     ANTO   GBPUSD       0.013555                 True
65   2022-01-06     ANTO   GBPUSD       0.013531                 True
82   2022-01-07     ANTO   GBPUSD       0.013590                 True
...         ...      ...      ...            ...                  ...
7634 2022-12-27      TGA   GBPUSD       0.011511                 True
7643 2022-12-28     ANTO   GBPUSD       0.011511                 True
7675 2022-12-28      TGA   GBPUSD       0.011511                 True
7684 2022-12-29     ANTO   GBPUSD       0.011511                 True
7715 2022-12-29      TGA   GBPUSD       0.011511                 True

[344 rows x 5 columns]

### Check for quantity change errors

This function, `check_quantity_trade_mismatch`, identifies discrepancies between the change in holding quantities (`Open Quantity` and `Close Quantity`) and the reported trades (`Traded Today`). It calculates the `Quantity Change` as the difference between `Close Quantity` and `Open Quantity`. Rows where the `Quantity Change` is non-zero but does not match the `Traded Today` value (within a small numerical tolerance, `atol=1e-6`) are flagged.

#### Formula:
$$
\text{Quantity Change} = \text{Close Quantity} - \text{Open Quantity}
$$

In [7]:
def check_quantity_trade_mismatch(data):
    data['Quantity Change'] = data['Close Quantity'] - data['Open Quantity']
    quantity_flags = data[
        (data['Quantity Change'] != 0) & 
        (~np.isclose(data['Traded Today'], data['Quantity Change'], atol=1e-6))
    ]
    return quantity_flags[['Date', 'P_Ticker', 'Open Quantity', 'Close Quantity', 'Quantity Change', 'Traded Today']]

# Example usage
quantity_flags = check_quantity_trade_mismatch(data)
quantity_flags

Empty DataFrame
Columns: [Date, P_Ticker, Open Quantity, Close Quantity, Quantity Change, Traded Today]
Index: []

### Check Prices where changes are greater than 10 percent compared to yesterday's price

This cell defines a function `check_inconsistent_prices` to identify rows in the dataset where the percentage change in price exceeds a specified threshold (10% in this case). It calculates the price change and the percentage change relative to the previous day's price, flags rows with significant changes, and returns a DataFrame containing the flagged rows.

#### Formula:
$$
\text{Price Change Percentage} = \frac{\text{Price} - \text{Price Yesterday}}{\text{Price Yesterday}}
$$


In [8]:
def check_inconsistent_prices(data):
    data['Price Change'] = data['Price'] - data['Price Yesterday']
    data['Price Change Percentage'] = data['Price Change'] / data['Price Yesterday']
    price_flags = data[np.abs(data['Price Change Percentage']) > 0.1]  # Flagging >10% change
    return price_flags[['Date', 'P_Ticker', 'Price', 'Price Yesterday', 'Price Change Percentage']]

# Example usage
price_flags = check_inconsistent_prices(data)
price_flags

Date P_Ticker       Price  Price Yesterday  Price Change Percentage
33   2022-01-04       SE  197.835007       223.184998                -0.113583
44   2022-01-05      AMC   22.745001        25.495001                -0.107864
49   2022-01-05      GME  129.335007       148.604996                -0.129673
53   2022-01-05     RIVN   90.019997       101.434998                -0.112535
139  2022-01-12      TMV   13.475000        11.672500                 0.154423
...         ...      ...         ...              ...                      ...
7366 2022-12-19     MARA    3.870000         4.720000                -0.180085
7374 2022-12-19     PALL  154.250000       177.919998                -0.133037
7619 2022-12-27     MARA    3.240000         3.620000                -0.104972
7636 2022-12-27     TSLA  109.099998       123.150002                -0.114089
7689 2022-12-29     CVNA    4.450000         3.830000                 0.161880

[184 rows x 5 columns]

### Similarly for Trade Price

This section focuses on identifying discrepancies in trade price changes. The function flags rows where the percentage change in trade price exceeds a specified threshold (e.g., 10%).

#### Formula:
$$
\text{Trade Price Change Percentage} = \frac{\text{Trade Price} - \text{Price}}{\text{Price}}
$$

In [9]:
def check_trade_price_vs_price_today(data):
    data['Trade Price Change Percentage'] = np.abs(data['Trade Price'] - data['Price']) / data['Price']
    trade_price_flags = data[data['Trade Price Change Percentage'] > 0.1]  # Flagging >10% change
    return trade_price_flags[['Date', 'P_Ticker', 'Price', 'Trade Price', 'Trade Price Change Percentage']]

# Example usage
trade_price_flags = check_trade_price_vs_price_today(data)
trade_price_flags

Date P_Ticker       Price  Trade Price  \
710  2022-03-02      RSX    7.190000     8.130000   
806  2022-03-07     RIVN   42.430000    48.000000   
807  2022-03-08   CLB_US   35.040001    31.408756   
844  2022-03-09     HSBK    9.445000     8.500000   
845  2022-03-09      KAP   26.549999    32.841137   
846  2022-03-09     KSPI   48.599998    39.650000   
1781 2022-04-21      NXE    6.010000     5.401450   
2054 2022-05-09     MSTR  219.255005   253.000000   
2105 2022-05-11     MSTR  168.244995   202.000000   
2482 2022-06-01     CVNA   26.495001    29.647139   
3078 2022-06-28     CVNA   24.735001    28.430000   
4179 2022-08-22      AMC   10.460000    12.600000   
4261 2022-08-24      AMC    9.580000     7.020000   
5022 2022-09-15   SSL_CA    7.370000     8.276563   
5357 2022-09-29      VNA   26.365000    21.320000   
5382 2022-09-30      VNA   26.365000    21.260000   
5383 2022-10-03      DBK    8.325500     7.362181   
5391 2022-10-03     GLTL   43.720001    48.800119   
5414 2022-10-04       LI   18.527500    25.501131   
5495 2022-10-07     GLTL   39.669998    46.848140   
5496 2022-10-07       LI   17.795000    20.306456   
5519 2022-10-07      VNA   26.365000    21.101522   
5593 2022-10-13     MARA   11.220000     9.530000   
5683 2022-10-17      VNA   26.365000    21.206130   
5860 2022-10-24      VNA   26.365000    21.134004   
5892 2022-10-25      VNA   26.365000    22.210000   
6007 2022-10-28      VNA   26.365000    22.484333   
6008 2022-10-31      CLF   16.809999    13.081161   
7213 2022-12-14     EXS1  111.290001   123.712890   
7291 2022-12-16     MARA    4.720000     3.980236   
7292 2022-12-16       SE   64.980003    55.503369   

      Trade Price Change Percentage  
710                        0.130737  
806                        0.131275  
807                        0.103631  
844                        0.100053  
845                        0.236954  
846                        0.184156  
1781                       0.101256  
2054                       0.153908  
2105                       0.200630  
2482                       0.118971  
3078                       0.149383  
4179                       0.204589  
4261                       0.267223  
5022                       0.123007  
5357                       0.191352  
5382                       0.193628  
5383                       0.115707  
5391                       0.116197  
5414                       0.376393  
5495                       0.180946  
5496                       0.141133  
5519                       0.199639  
5593                       0.150624  
5683                       0.195671  
5860                       0.198407  
5892                       0.157595  
6007                       0.147190  
6008                       0.221823  
7213                       0.111626  
7291                       0.156730  
7292                       0.145839

### Calculate and Check Profit Errors

This cell defines a function `calculate_and_check_profit_errors` to calculate the Dollar Profit and Loss (PnL) based on each row of data and including if a trade has been carried out and compare it with the existing Dollar PnL values. It flags rows where the calculated Dollar PnL does not match the existing Dollar PnL within a specified tolerance.

#### Formula:
1. **Dollar PnL Calculation**:
    - If `Open Quantity` equals `Close Quantity`:
      $$
      \text{Dollar PnL} = (\text{Price} - \text{Price Yesterday}) \times \text{Close Quantity} \times \text{Exchange Rate}
      $$
    - If `Close Quantity` is greater than `Open Quantity`:
      $$
      \text{Dollar PnL} = \text{Open Quantity} \times (\text{Price} - \text{Price Yesterday}) \times \text{Exchange Rate} - (\text{Close Quantity} - \text{Open Quantity}) \times (\text{Trade Price} - \text{Price}) \times \text{Exchange Rate}
      $$
    - If `Close Quantity` is less than `Open Quantity`:
      $$
      \text{Dollar PnL} = (\text{Open Quantity} - \text{Close Quantity}) \times (\text{Trade Price} - \text{Price}) \times \text{Exchange Rate} + \text{Open Quantity} \times (\text{Price} - \text{Price Yesterday}) \times \text{Exchange Rate}
      $$

In [10]:
def calculate_and_check_profit_errors(data):
    # Calculate Dollar PnL
    data['Calculated Dollar PnL'] = np.where(
        data['Open Quantity'] == data['Close Quantity'],
        (data['Price'] - data['Price Yesterday']) * data['Close Quantity'] * data['Exchange Rate'],
        np.where(
            data['Close Quantity'] > data['Open Quantity'],
            data['Open Quantity'] * (data['Price'] - data['Price Yesterday']) * data['Exchange Rate'] - 
            (data['Close Quantity'] - data['Open Quantity']) * (data['Trade Price'] - data['Price']) * data['Exchange Rate'],
            (data['Open Quantity'] - data['Close Quantity']) * (data['Trade Price'] - data['Price']) * data['Exchange Rate'] +
            data['Open Quantity'] * (data['Price'] - data['Price Yesterday']) * data['Exchange Rate']
        )
    )

    hidden_flags = data[
        ~np.isclose(data['Dollar PnL'], data['Calculated Dollar PnL'], atol=1e-6)
    ]
    
    return hidden_flags[['Date', 'P_Ticker', 'Dollar PnL', 'Calculated Dollar PnL']]

# Example usage
pnl_flags = calculate_and_check_profit_errors(data)
pnl_flags

Date P_Ticker  Dollar PnL  Calculated Dollar PnL
4221 2022-08-23      AMC  668.086395              641.89945

### Calculate Value in USD and Exchange Rate Errors

This function, `check_currency_conversion_errors`, identifies discrepancies in currency conversion calculations within a dataset. It computes the "Calculated Value in USD" for each row by multiplying the `Price`, `Exchange Rate`, and `Close Quantity`. It then compares this calculated value with the existing `Value in USD` column using a tolerance (`atol=1e-6`) to account for minor numerical differences.

#### Formula:
$$
\text{Calculated Value in USD} = \text{Price} \times \text{Exchange Rate} \times \text{Close Quantity}
$$

In [11]:
def check_currency_conversion_errors(data):
    data['Calculated Value in USD (Currency Check)'] = data['Price'] * data['Exchange Rate'] * data['Close Quantity']
    data['Currency Conversion Error'] = ~np.isclose(data['Value in USD'], data['Calculated Value in USD (Currency Check)'], atol=1e-6)
    
    # Filter rows with currency conversion errors
    currency_conversion_errors = data[data['Currency Conversion Error']]
    return currency_conversion_errors[['Date', 'P_Ticker', 'Value in USD', 'Calculated Value in USD (Currency Check)']]

currency_conversion_errors = check_currency_conversion_errors(data)
currency_conversion_errors

Date P_Ticker   Value in USD  \
371  2022-02-04     EXW1 -472104.129890   
762  2022-03-04      RHM  121400.221237   
792  2022-03-07      RHM  126674.955045   
988  2022-03-14      PFE   25891.669395   
1853 2022-04-27   ETFMIB -238910.180289   
...         ...      ...            ...   
7658 2022-12-28     KSPI   58604.001389   
7684 2022-12-29     ANTO  -66676.994958   
7697 2022-12-29     HSBK   40267.303764   
7698 2022-12-29      KAP   38692.500000   
7699 2022-12-29     KSPI   58604.001389   

      Calculated Value in USD (Currency Check)  
371                                   0.000000  
762                              -94936.957752  
792                              -99061.885825  
988                               28860.340000  
1853                                  0.000000  
...                                        ...  
7658                              65065.000000  
7684                             -86869.796171  
7697                              47778.200772  
7698                              42333.278988  
7699                              65520.000000  

[231 rows x 4 columns]

### Validate Opening and Closing Weights

Given the identified discrepancies in the `Value in USD`, it is crucial to validate the accuracy of the `Opening Weights` and `Closing Weights` as they are directly dependent on the total `Value in USD`.

#### Formulas:
1. **Opening Weights**:  
    $$
    \text{Calculated Opening Weights} = \frac{\text{Price Yesterday} \times \text{Exchange Rate} \times \text{Open Quantity}}{\text{NAV Yesterday}}
    $$

2. **Closing Weights**:  
    $$
    \text{Calculated Closing Weights} = \frac{\text{Price} \times \text{Close Quantity} \times \text{Exchange Rate}}{\text{Calculated NAV}}
    $$

In [12]:
def check_weights_errors(data):
    # Calculate Opening Weights
    data['Calculated Opening Weights'] = (data['Price Yesterday'] * data['Exchange Rate'] * data['Open Quantity']) / data['NAV Yesterday']
    data['Opening Weights Error'] = ~np.isclose(data['Opening Weights'], data['Calculated Opening Weights'], atol=1e-6)

    # Calculate Closing Weights
    data['Calculated Closing Weights'] = (data['Price'] * data['Close Quantity'] * data['Exchange Rate']) / data['Calculated NAV']
    data['Closing Weights Error'] = ~np.isclose(data['Closing Weights'], data['Calculated Closing Weights'], atol=1e-6)

    # Filter rows with errors
    weights_errors = data[data['Opening Weights Error'] | data['Closing Weights Error']]
    return weights_errors[['Date', 'P_Ticker', 'Opening Weights', 'Calculated Opening Weights', 
                           'Opening Weights Error', 'Closing Weights', 'Calculated Closing Weights', 
                           'Closing Weights Error']]

# Example usage
weights_errors = check_weights_errors(data)
weights_errors

Empty DataFrame
Columns: [Date, P_Ticker, Opening Weights, Calculated Opening Weights, Opening Weights Error, Closing Weights, Calculated Closing Weights, Closing Weights Error]
Index: []

### Check Market Cap Errors

This function, `check_market_cap_errors`, identifies discrepancies in market capitalization calculations within a dataset. It computes the "Calculated Market Cap" for each row by multiplying the `SharesOut`, `Price`, and `Exchange Rate`. It then compares this calculated value with the existing `Market_Cap` column using a tolerance (`atol=1e-6`) to account for minor numerical differences.

#### Formula:
$$
\text{Calculated Market Cap} = \text{SharesOut} \times \text{Price} \times \text{Exchange Rate}
$$

In [14]:
def check_market_cap_errors(data):
    # Calculate Market Cap
    data['Calculated Market Cap'] = data['SharesOut'] * data['Price'] * data['Exchange Rate']
    data['Market Cap Error'] = ~np.isclose(data['Market_Cap'], data['Calculated Market Cap'], atol=1e-6)
    
    # Filter rows with market cap errors
    market_cap_errors = data[data['Market Cap Error']]
    return market_cap_errors[['Date', 'P_Ticker', 'Market_Cap', 'Calculated Market Cap', 'Market Cap Error']]

# Example usage
market_cap_errors = check_market_cap_errors(data)
market_cap_errors

Empty DataFrame
Columns: [Date, P_Ticker, Market_Cap, Calculated Market Cap, Market Cap Error]
Index: []

### Check Stock Movement and Performance

This function, `check_stock_movement_and_performance`, recalculates the stock movement and performance contribution for each row in the dataset. It compares the recalculated values with the existing ones to identify discrepancies. The function flags rows where the calculated stock movement or performance contribution does not match the existing values within a specified tolerance (`atol=1e-6`). 

#### Formulas:
1. **Stock Movement**:  
    $$
    \text{Calculated Stock Movement} = \frac{\text{Price} - \text{Price Yesterday}}{\text{Price Yesterday}}
    $$

2. **Performance Contribution**:  
    $$
    \text{Calculated Performance Contribution} = \text{Close Quantity} \times \text{Price} \times \text{Exchange Rate} \times \text{Calculated Stock Movement}
    $$

In [17]:
def check_stock_movement_and_performance(data):
    # Recalculate Stock Movement and Performance Contribution
    data['Calculated Stock Movement'] = (data['Price'] - data['Price Yesterday']) / data['Price Yesterday']
    data['Calculated Performance Contribution'] = data['Close Quantity'] * data['Price'] * data['Exchange Rate'] * data['Calculated Stock Movement']
    
    # Check for discrepancies
    data['Stock Movement Error'] = ~np.isclose(data['Stock Movement'], data['Calculated Stock Movement'], atol=1e-6)
    data['Performance Contribution Error'] = ~np.isclose(data['Performance Contribution'], data['Calculated Performance Contribution'], atol=1e-6)
    
    # Filter rows with errors
    stock_performance_errors = data[
        data['Stock Movement Error'] | data['Performance Contribution Error']
    ]
    
    return stock_performance_errors[['Date', 'P_Ticker', 'Stock Movement', 'Calculated Stock Movement', 
                                     'Stock Movement Error', 'Performance Contribution', 
                                     'Calculated Performance Contribution', 'Performance Contribution Error']]

# Example usage
stock_performance_errors = check_stock_movement_and_performance(data)
stock_performance_errors

Empty DataFrame
Columns: [Date, P_Ticker, Stock Movement, Calculated Stock Movement, Stock Movement Error, Performance Contribution, Calculated Performance Contribution, Performance Contribution Error]
Index: []

### Check Trade Day Move and Weights

This function, `check_trade_day_move_and_weights`, recalculates the trade day move, trade weight, and return adjustments for rows where trades occurred (`Traded Today` is not NaN). It compares the recalculated values with the existing ones to identify discrepancies. The function flags rows where the calculated trade day move, trade weight, or return adjustments do not match the existing values within a specified tolerance (`atol=1e-6`).

#### Formulas:
1. **Trade Day Move**:  
    $$
    \text{Calculated Trade Day Move} = \frac{\text{Price} - \text{Trade Price}}{\text{Trade Price}}
    $$

2. **Trade Weight**:  
    $$
    \text{Calculated Trade Weight} = \frac{\text{Traded Today} \times \text{Trade Price} \times \text{Exchange Rate}}{\text{NAV Yesterday}}
    $$

3. **Return Adjustments**:  
    $$
    \text{Calculated Return Adjustments} = \text{Calculated Trade Weight} \times \text{Calculated Trade Day Move}
    $$

In [26]:
def check_trade_day_move_and_weights(data):
    # Perform calculations only for rows where 'Traded Today' is not NaN
    data.loc[data['Traded Today'].notna(), 'Calculated Trade Day Move'] = (
        (data['Price'] - data['Trade Price']) / data['Trade Price']
    )
    data.loc[data['Traded Today'].notna(), 'Calculated Trade Weight'] = (
        (data['Traded Today'] * data['Trade Price'] * data['Exchange Rate']) / data['NAV Yesterday']
    )
    data.loc[data['Traded Today'].notna(), 'Calculated Return Adjustments'] = (
        data['Calculated Trade Weight'] * data['Calculated Trade Day Move']
    )
    
    # Check for discrepancies
    data['Trade Day Move Error'] = ~np.isclose(data['Trade Day Move'], data['Calculated Trade Day Move'], atol=1e-6)
    data['Trade Weight Error'] = ~np.isclose(data['Trade Weight'], data['Calculated Trade Weight'], atol=1e-6)
    data['Return Adjustment Error'] = ~np.isclose(data['Return Adjustments'], data['Calculated Return Adjustments'], atol=1e-6)
    
    # Filter rows with errors and where 'Traded Today' is not NaN
    trade_day_move_and_weights_errors = data.loc[
        data['Traded Today'].notna() & 
        (data['Trade Day Move Error'] | data['Trade Weight Error'] | data['Return Adjustment Error'])
    ]
    
    return trade_day_move_and_weights_errors[['Date', 'P_Ticker', 'Trade Day Move', 'Calculated Trade Day Move', 
                                               'Trade Weight', 'Calculated Trade Weight', 
                                               'Return Adjustments', 
                                              'Calculated Return Adjustments']]

# Example usage
trade_day_move_and_weights_errors = check_trade_day_move_and_weights(data)
trade_day_move_and_weights_errors

Date P_Ticker  Trade Day Move  Calculated Trade Day Move  \
4221 2022-08-23      AMC       -0.048756                  -0.048756   

      Trade Weight  Calculated Trade Weight  Return Adjustments  \
4221     -0.010426                -0.010017            0.000508   

      Calculated Return Adjustments  
4221                       0.000488

In [27]:
all_flags = pd.concat([exchange_rate_errors,quantity_flags, price_flags, trade_price_flags, pnl_flags, weights_errors, currency_conversion_errors, market_cap_errors, stock_performance_errors,trade_day_move_and_weights_errors], axis=0)
all_flags

Date P_Ticker Currency  Exchange Rate Exchange Rate Error  \
6    2022-01-03     ANTO   GBPUSD       0.013480                True   
27   2022-01-04     ANTO   GBPUSD       0.013527                True   
46   2022-01-05     ANTO   GBPUSD       0.013555                True   
65   2022-01-06     ANTO   GBPUSD       0.013531                True   
82   2022-01-07     ANTO   GBPUSD       0.013590                True   
...         ...      ...      ...            ...                 ...   
7684 2022-12-29     ANTO      NaN            NaN                 NaN   
7697 2022-12-29     HSBK      NaN            NaN                 NaN   
7698 2022-12-29      KAP      NaN            NaN                 NaN   
7699 2022-12-29     KSPI      NaN            NaN                 NaN   
4221 2022-08-23      AMC      NaN            NaN                 NaN   

      Open Quantity  Close Quantity  Quantity Change  Traded Today  Price  \
6               NaN             NaN              NaN           NaN    NaN   
27              NaN             NaN              NaN           NaN    NaN   
46              NaN             NaN              NaN           NaN    NaN   
65              NaN             NaN              NaN           NaN    NaN   
82              NaN             NaN              NaN           NaN    NaN   
...             ...             ...              ...           ...    ...   
7684            NaN             NaN              NaN           NaN    NaN   
7697            NaN             NaN              NaN           NaN    NaN   
7698            NaN             NaN              NaN           NaN    NaN   
7699            NaN             NaN              NaN           NaN    NaN   
4221            NaN             NaN              NaN           NaN    NaN   

      Price Yesterday  Price Change Percentage  Trade Price  \
6                 NaN                      NaN          NaN   
27                NaN                      NaN          NaN   
46                NaN                      NaN          NaN   
65                NaN                      NaN          NaN   
82                NaN                      NaN          NaN   
...               ...                      ...          ...   
7684              NaN                      NaN          NaN   
7697              NaN                      NaN          NaN   
7698              NaN                      NaN          NaN   
7699              NaN                      NaN          NaN   
4221              NaN                      NaN          NaN   

      Trade Price Change Percentage  Dollar PnL  Calculated Dollar PnL  \
6                               NaN         NaN                    NaN   
27                              NaN         NaN                    NaN   
46                              NaN         NaN                    NaN   
65                              NaN         NaN                    NaN   
82                              NaN         NaN                    NaN   
...                             ...         ...                    ...   
7684                            NaN         NaN                    NaN   
7697                            NaN         NaN                    NaN   
7698                            NaN         NaN                    NaN   
7699                            NaN         NaN                    NaN   
4221                            NaN         NaN                    NaN   

      Opening Weights  Calculated Opening Weights Opening Weights Error  \
6                 NaN                         NaN                   NaN   
27                NaN                         NaN                   NaN   
46                NaN                         NaN                   NaN   
65                NaN                         NaN                   NaN   
82                NaN                         NaN                   NaN   
...               ...                         ...                   ...   
7684              NaN                         NaN 

There are 792 flags that could potentially result in errors and thus flagged 